src/qiskit_qasm3_import/exceptions.py

```diff
@@ -8,9 +8,23 @@ class ConversionError(Exception):
     :class:`~qiskit.circuit.QuantumCircuit`.  This is often due to OpenQASM 3 constructs that have
     no equivalent in Qiskit."""
 
+    def __init__(self, message, node: Optional[ast.QASMNode] = None):
+        if node is not None and node.span is not None:
+            message = f"{node.span.start_line},{node.span.start_column}: {message}"
+        self.message = message
+        super().__init__(message)
+
 
 def raise_from_node(node: ast.QASMNode, message: str) -> NoReturn:
     """Raise a :exc:`.ConversionError` caused by the given `node`."""
-    if node.span is not None:
-        message = f"{node.span.start_line},{node.span.start_column}: {message}"
-    raise ConversionError(message)
+    raise ConversionError(message, node)
+
+
+class PhysicalQubitInGateError(ConversionError):
+    def __init__(self, name: str, node=None):
+        self.name = name
+        msg = (
+            f"Illegal qubit reference '{name}'. References to hardware "
+            "qubits not allowed in gate definitions."
+        )
+        super().__init__(msg, node)
```

Comment:
I'm not a fan of doing this in a separate class, mostly because in an ideal world we wouldn't be raising exceptions at all, we'd be appending the messages to a diagnostic stream and then only finally raising an exception (library) / returning a non-zero error code (application).  There's also the weird API where a user could ask "why is _this_ error separate and not _that_ error?"; it may well be subjective, but imo we should only use separate exception classes when we actively expect users to want to behave differently upon `try/except`'ing them.

I think you can achieve what you want (unified error message) without the new class if you use a little helper function to encapsulate the string.

Reply:
Using a separate method only for this one case is only because I didn't want to take the time to redesign everything.

I agree 100% that exceptions are not a priori the correct way to handle these errors in translation. In practice, doing something other than exiting immediately would be more complex.

I did consider using a scheme other than exceptions to make this more clean; some kind of tag. But the exception machinery is tempting because most of the work is already done for us. I'll try something else.